### ISSUE TITLE
Omitting semicolons is not future-proof

### ISSUE THREAD

**Issue by bakkot** [feross/standard#581]
Consider for example the [public fields proposal](https://jeffmo.github.io/es-class-public-fields/).

Under it, this is valid code:

```js
class A {
  get;
  set;
}
```

and this is a syntax error:
```js
class A {
  get
  set
}
```

You could add a rule which says that class fields must always start or end with a semicolon (or some other, even more special-case-y rule), but it's more overhead to remember.

I don't mean to rehash the great semicolon debate; I just wanted to point out that this style is only going to become harder and harder to follow as the language evolves.

(Incidentally, Babel's parsing of class properties is fairly broken, so don't rely on it as an oracle here.)

--

```js
class A {
  x
  y
}
```
is legal; it's just 'get', 'set', 'static', and similar names which have this issue. ASI will not put a semicolon in the second snippet above because in class bodies an identifier such as 'set' is in fact a legal token to follow 'get' without an intervening semicolon.

**Comment by timoxley** [comment]
I think you're throwing the baby out with the bath water. Cases like get/set/static would fall in the realm of "necessary' semicolons, like those in for loops, and appears easy to identify. Perhaps we need to add a few more semicolons in, no drama. Standard isn't so zealous about "no semicolons" that it's going to disallow/break valid JavaScript constructs just because they require semicolons. Standard simply enforces no *unnecessary* semicolons. Adding new rules to handle new syntax is an inevitability. I don't see any problem here.

**Comment by yoshuawuyts** [comment]
@bakkot thanks for the feedback. You make a valid point, but like Tim said we'll add new rules when new syntax is introduced.

Closing because there's not really anything to be done here now

**Comment by bakkot** [comment]
For posterity, this is related to https://github.com/feross/standard/issues/372.

**Comment by jmeas** [comment]
Cross-referencing https://github.com/tc39/ecma262/pull/1062 , which seems related.

tl;dr, TC39 discourages the practice of relying on ASI.

**Comment by bcomnes** [comment]
@jmeas I responded https://github.com/tc39/ecma262/pull/1062#issuecomment-357036281

I think they could improve their neutrality and suggested a re-wording on the issue to more accurately and fairly represent the issue.
